src/main/kotlin/org/jetbrains/research/ml/codetracker/server/TrackerQueryExecutor.kt

```diff
@@ -0,0 +1,105 @@
+
+    private var activityTrackerKey: String? = null
+
+    init {
+        initActivityTrackerInfo()
+    }
+
+    private fun initActivityTrackerInfo() {
+        val currentUrl = URL(baseUrl + "activity-tracker-item")
+        logger.info("${Plugin.PLUGIN_ID}: ...generating activity tracker key")
+
+        val request = Request.Builder().url(currentUrl).post(
+            ByteArray(0)
+                .toRequestBody(null, 0, 0)
+        ).build()
+        activityTrackerKey = executeQuery(request).get()?.let { it.body?.string() }
+    }
+
+    private fun createTrackerRequestBody(
+        fileFieldName: String, file: File,
+        toAddActivityTrackerKey: Boolean = false
+    ): MultipartBody.Builder {
+        val requestBody = MultipartBody.Builder()
+            .setType(MultipartBody.FORM)
+            .addFormDataPart(
+                fileFieldName, file.name,
+                file.asRequestBody("text/csv".toMediaType())
+            )
+        if (toAddActivityTrackerKey) {
+            activityTrackerKey?.let { requestBody.addFormDataPart("activityTrackerKey", it) }
+        }
+        return requestBody
+    }
+
+    private fun sendActivityTrackerData() {
+        val file = File(activityTrackerPath)
+        if (file.exists()) {
+            val currentUrl = baseUrl + "activity-tracker-item/" + activityTrackerKey
+            logger.info("${Plugin.PLUGIN_ID}: ...sending file ${file.name}")
+            executeQuery(
+                Request.Builder()
+                    .url(currentUrl)
+                    .put(
+                        createTrackerRequestBody(
+                            ACTIVITY_TRACKER_FILE_FIELD,
+                            file
+                        ).build()
+                    )
+                    .build()
+            )
+        } else {
+            logger.info("${Plugin.PLUGIN_ID}: activity-tracker file doesn't exist")
+        }
+    }
+
+    fun sendCodeTrackerData(file: File, deleteAfter: () -> Boolean, postActivity: () -> Unit) {
+        val currentUrl = baseUrl + "org.jetbrains.research.ml.codetracker.data-item"
+        logger.info("${Plugin.PLUGIN_ID}: ...sending file ${file.name}")
+        val future = executeQuery(
+            Request.Builder()
+                .url(currentUrl)
+                .post(
+                    createTrackerRequestBody(
+                        CODE_TRACKER_FILE_FIELD,
+                        file,
+                        activityTrackerKey != null
+                    ).build()
+                )
+                .build()
+        )
+        if (isSuccess(future.get())) {
+            if (deleteAfter()) {
+                logger.info("${Plugin.PLUGIN_ID}: delete file ${file.name}")
+                file.delete()
+            }
+            if (activityTrackerKey != null) {
+                sendActivityTrackerData()
+            }
+        }
+        postActivity()
+    }
+
+    fun sendCodeTrackerData(file: File) {
```

Comment:
Why do we need this function? Isn't it simpler to set default parameters to the upper function?